File: exercises/task_scheduler_621.py
from heapq import heappush, heappop
from typing import List

# MaxHeap
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        freq = [0] * 27
        maxHeap = []
        for ch in tasks:
            freq[ord(ch) - ord("A")] += 1
        for f in freq:
            if f > 0: heappush(maxHeap, -f)
        time = 0
        while len(maxHeap) > 0:
            addBack = []
            for _ in range(n+1):
                if len(maxHeap) > 0:
                    count = -heappop(maxHeap)
                    count -= 1
                    if count > 0:
                        addBack.append(count)
                time += 1
                if len(maxHeap) == 0 and len(addBack) == 0:
                    break
            for count in addBack:
                heappush(maxHeap, -count)
        return time

# Math Solution
# class Solution:
#     def leastInterval(self, tasks: List[str], n: int) -> int:
#         freq = [0] * 27
#         for ch in tasks:
#             freq[ord(ch) - ord("A")] += 1
#         maxFreq, maxCount = 0, 0
#         for f in freq:
#             if f == 0: continue
#             if f > maxFreq:
#                 maxFreq = f
#                 maxCount = 1
#             elif f == maxFreq:
#                 maxCount += 1
#         maxTasksCycles = maxFreq - 1
#         gapsBetweenMaxTasks = n - (maxCount - 1)
#         emptySlots = maxTasksCycles * gapsBetweenMaxTasks
#         availableTasks = len(tasks) - (maxFreq * maxCount)
#         idleSlots = max(0, emptySlots - availableTasks)
#         return len(tasks) + idleSlots

File: exercises/task_scheduler_621.java
import java.util.LinkedList;
import java.util.PriorityQueue;

// Math Algorithm Solution
class Solution {
	public int leastInterval(char[] tasks, int n) {
		// Count frequencies
		int[] freq = new int[26];
		for (char task : tasks) {
			freq[task - 'A']++;
		}
		// Find max frequency and count of tasks with max frequency
		int maxFreq = 0, maxCount = 0;
		for (int f : freq) {
			if (f > maxFreq) {
				maxFreq = f;
				maxCount = 1;
			} else if (f == maxFreq) {
				maxCount++;
			}
		}
		int maxTasksCycles = maxFreq - 1;
		int gapsBetweenCycles = n - (maxCount - 1);
		int emptySlots = maxTasksCycles * gapsBetweenCycles;
		int availableTasks = tasks.length - maxFreq * maxCount;
		int idleSlots = Math.max(0, emptySlots - availableTasks);
		return tasks.length + idleSlots;
	}
}

// Priority Queue Solution (inefficient)
//class Solution {
//    public int leastInterval(char[] tasks, int n) {
//        int[] freq = new int[26];
//        for (char c : tasks) {
//            freq[c - 'A']++;
//        }
//
//        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> b - a);
//        for (int f : freq) {
//            if (f > 0) pq.offer(f);
//        }
//
//        Queue<Pair<Integer, Integer>> q = new LinkedList<>();
//        int time = 0;
//
//        while (!pq.isEmpty() || !q.isEmpty()) {
//            time++;
//
//            if (!pq.isEmpty()) {
//                int cnt = pq.poll() - 1;
//
//                if (cnt != 0) {
//                    q.offer(new Pair(cnt, time + n));
//                }
//            }
//
//            if (!q.isEmpty() && q.peek().getValue() == time) {
//                pq.offer(q.poll().getKey());
//            }
//        }
//
//        return time;
//    }
//}
